prover/docs/04_flow.md

```diff
@@ -0,0 +1,209 @@
+# Prover flow
+
+In this section, we're going to learn what stages does the proof generation process have. It's a complex process, so
+we'll be looking at it from four perspectives:
+
+- Core<->Prover subsystem interactions.
+- Core side of workflow.
+- Prover pipeline.
+- Actual proof generation.
+- Infrastructure distribution.
+
+After that, we will touch on how this flow is mapped on the actual production infrastructure.
+
+## Core <-> Prover subsystem interactions
+
+Core and prover subsystem are built in such a way that they are mostly isolated from each other. Each side has its own
+database and GCS buckets, and both have "gateway" components they use for interaction.
+
+The only exception here is the `house_keeper`: it's a component that exists as a part of the server, it's main purpose
+is to manage jobs in the prover workspace, but at the same time it has access to both core and prover databases.
+
+Otherwise, the interaction between subsystems can be expressed as follows:
+
+```mermaid
+sequenceDiagram
+  participant C as Core
+  participant P as Prover
+
+  loop In parallel, for each batch
+    P-->>+C: Get a job to prove
+    C->>-P: Unproven batch
+    P->>P: Calculate proof
+    P->>C: Submit proof
+  end
+```
+
+Core exposes an API, and Prover repeatedly polls this API, fetching new jobs and submitting ready proofs.
+
+## Core side of workflow
+
+Despite the fact that the prover is isolated from the core, the core has multiple components specifically designed to
+prepare _inputs_ for proving.
+
+The following diagram shows what happens under the hood when the prover subsystem requests a new job:
+
+```mermaid
+sequenceDiagram
+  box Core
+  participant Ob as GCS
+  participant DB as Core database
+  participant API as Proof data handler
+  end
+  participant P as Prover
+  P-->>+API: Get a job
+  API-->>DB: Lock a suitable job
+  DB->>API: Job is marked as "picked_up"
+  API-->>Ob: Fetch BWIP data
+  Ob->>API: Return BWIP data
+  API-->>Ob: Fetch Merkle Tree data
+  Ob->>API: Return Merkle Tree data
+  API-->>DB: Fetch batch metadata
+  DB->>API: Return batch metadata
+  API->>-P: Return a job
+```
+
+First of all, `proof_data_handler` will check if all the data required for the proof generation is already prepared by
+the core. If so, it will lock the job so that it's not assigned twice, and will fetch required information from multiple
+sources. Then this data is given to the prover together with the batch number.
+
+## Prover pipeline
+
+Once job is received by the prover, it has to go through several different stages:
+
+```mermaid
```

Comment:
As a person who only recently started to learn about prover workspace, I find this information extremely useful, because it gives you a mental model that helps you to understand things better as you learn. Without it you have a ton of interdependent jobs which -- as you said -- have different pipeline properties w.r.t. parallelism.

I've added a note that it's only a mental model though.